osu.Desktop/DiscordRichPresenceClient.cs

```diff
@@ -0,0 +1,122 @@
+
+            user.ValueChanged += usr =>
+            {
+                usr.NewValue.Activity.ValueChanged += activity => updateStatus(user.Value.Status.Value, activity.NewValue);
+                usr.NewValue.Status.ValueChanged += status => updateStatus(status.NewValue, user.Value.Activity.Value);
+            };
+
+            user.TriggerChange();
+
+            enableLogging();
+
+            client.Initialize();
+        }
+
+        private void enableLogging()
+        {
+            client.OnReady += (_, __) => Logger.Log("Discord RPC Client ready.", LoggingTarget.Network, LogLevel.Debug);
+            client.OnError += (_, e) => Logger.Log($"An error occurred with Discord RPC Client : {e.Message}", LoggingTarget.Network, LogLevel.Debug);
+            client.OnConnectionFailed += (_, e) => Logger.Log("Discord RPC Client failed to initialize : is discord running ?", LoggingTarget.Network, LogLevel.Debug);
+            client.OnPresenceUpdate += (_, __) => Logger.Log("Updated Discord Rich Presence", LoggingTarget.Network, LogLevel.Debug);
+        }
+
+        private void updateStatus(UserStatus st, UserActivity a)
+        {
+            var presence = defaultPresence(st is UserStatusOnline ? a?.Status : st.Message);
+
+            if (!(st is UserStatusOnline)) //don't update the presence any further if the current user status is DND / Offline & simply return with the default presence
+            {
+                client.SetPresence(presence);
+                return;
+            }
+
+            switch (a)
+            {
+                case UserActivitySoloGame game:
+                    presence.State = $"{game.Beatmap.Metadata.Artist} - {game.Beatmap.Metadata.Title} [{game.Beatmap.Version}]";
+                    setPresenceGamemode(game.Ruleset, presence);
+                    break;
+
+                case UserActivityEditing editing:
+                    presence.State = $"{editing.Beatmap.Metadata.Artist} - {editing.Beatmap.Metadata.Title} " + (!string.IsNullOrEmpty(editing.Beatmap.Version) ? $"[{editing.Beatmap.Version}]" : "");
+                    presence.Assets.SmallImageKey = "edit";
+                    presence.Assets.SmallImageText = "editing";
+                    break;
+            }
+
+            client.SetPresence(presence);
+        }
+
+        private void setPresenceGamemode(RulesetInfo ruleset, RichPresence presence)
+        {
+            switch (ruleset.ID)
+            {
+                case 0:
+                    presence.Assets.SmallImageKey = ruleset.ShortName;
+                    break;
+
+                case 1:
+                    presence.Assets.SmallImageKey = ruleset.ShortName;
+                    break;
+
+                case 2:
+                    presence.Assets.SmallImageKey = ruleset.ShortName;
+                    break;
+
+                case 3:
+                    presence.Assets.SmallImageKey = ruleset.ShortName;
+                    break;
+
+                default:
```

Comment:
can be shortened to 
```
case 0:
case 1:
case 2:
case 3:
    //logic
```
Either way it's probably pointless to use a switch-case for just 2 different outcomes now

Reply:
You're right, I haven't thought of that. Let's use a if statement instead then